### PR TITLE
Make python{2,3} version explicit

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change
Changes for issue #18185

###### Things done

- [ ] Tested using sandboxing
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change
- [ ] Tested compilation of all pkgs that depend on this change
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/btfs/default.nix [hunk 1]
```diff
@@ -20,7 +20,7 @@ stdenv.mkDerivation rec {
 
   preInstall = ''
     substituteInPlace scripts/btplay \
-      --replace "/usr/bin/env python" "${python}/bin/python"
+      --replace "/usr/bin/env python" "${python3}/bin/python"
```
Comment (FRidh) [inline pkgs/os-specific/linux/btfs/default.nix]: ${python.3.interpreter}

File: pkgs/development/python-modules/robotframework-python3/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{ stdenv, fetchPypi, buildPythonPackage, six }:
+
+with stdenv.lib;
+
+buildPythonPackage rec {
+  name  = "${pname}-${version}";
```
Comment (FRidh) [inline pkgs/development/python-modules/robotframework-python3/default.nix]: The `name` attribute is added by `buildPython*` and should therefore be removed.

File: pkgs/servers/matrix-synapse/default.nix [hunk 3]
```diff
@@ -1,6 +1,6 @@
-{ stdenv, pythonPackages, fetchurl, fetchFromGitHub }:
+{ lib, pkgs, stdenv, python2Packages, fetchurl, fetchFromGitHub, fetchpatch }:
```
Comment (FRidh) [inline pkgs/servers/matrix-synapse/default.nix]: Don't pass in `pkgs`. Unused as well.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/robotsuite/default.nix [hunk 4]
```diff
@@ -10,7 +15,13 @@ buildPythonPackage rec {
   };
 
   buildInputs = [ unittest2 ];
-  propagatedBuildInputs = [ robotframework lxml ];
+  propagatedBuildInputs = [ robotframework lxml ]
+    ++ stdenv.lib.optional isPy3k robotframework-python3;
```
Comment (dotlambda) [inline pkgs/development/python-modules/robotsuite/default.nix]: From https://pypi.org/project/robotframework-python3/:

> Compatible with Python 2.7

So I think we should use `robotframework-python3` in both cases. This way we can get rid of `postPatch`.

File: pkgs/tools/misc/asciinema/default.nix [hunk 5]
```diff
@@ -1,13 +1,11 @@
 { lib, python3Packages, fetchFromGitHub, glibcLocales }:
 
-let
-  pythonPackages = python3Packages;
-in pythonPackages.buildPythonApplication rec {
+python3Packages.buildPythonApplication rec {
   name = "asciinema-${version}";
```
Comment (dotlambda) [inline pkgs/tools/misc/asciinema/default.nix]: `pname`

File: pkgs/tools/networking/offlineimap/default.nix [hunk 6]
```diff
@@ -1,7 +1,7 @@
-{ stdenv, fetchFromGitHub, pythonPackages,
+{ stdenv, fetchFromGitHub, python2Packages,
   asciidoc, libxml2, libxslt, docbook_xsl }:
 
-pythonPackages.buildPythonApplication rec {
+python2Packages.buildPythonApplication rec {
   version = "7.2.1";
   name = "offlineimap-${version}";
   namePrefix = "";
```
Comment (dotlambda) [inline pkgs/tools/networking/offlineimap/default.nix]: `pname`, no `namePrefix`

File: pkgs/os-specific/linux/btfs/default.nix (hunk 1 above)
Comment (rnhmjoj) [reply pkgs/os-specific/linux/btfs/default.nix]: @FRidh Is it `${python3.interpreter}`, right?

File: pkgs/servers/matrix-synapse/default.nix (hunk 3 above)
Comment (rnhmjoj) [reply pkgs/servers/matrix-synapse/default.nix]: Sorry, I must have messed something up when rebasing.

File: pkgs/tools/networking/offlineimap/default.nix (hunk 6 above)
Comment (rnhmjoj) [reply pkgs/tools/networking/offlineimap/default.nix]: Fixed.

File: pkgs/tools/misc/asciinema/default.nix (hunk 5 above)
Comment (rnhmjoj) [reply pkgs/tools/misc/asciinema/default.nix]: Fixed.

File: pkgs/os-specific/linux/btfs/default.nix (hunk 1 above)
Comment (FRidh) [reply pkgs/os-specific/linux/btfs/default.nix]: ah, yes

File: pkgs/development/python-modules/robotsuite/default.nix (hunk 4 above)
Comment (rnhmjoj) [reply pkgs/development/python-modules/robotsuite/default.nix]: The package actually requires this (now outdated) fork for python3 because the mainline version was python2 only at the time. I'm not sure if we can use the latest robotframework for both python 2/3 builds.
I must have left the `postPatch` here after testing this. It does build but I don't know how  to properly test it.

Comment (dotlambda) [reply pkgs/development/python-modules/robotsuite/default.nix]: Ah, I see. Then we should simply patch `robotsuite`'s `setup.py`:
```
postPatch = ''
  substituteInPlace setup.py \
    --replace robotframework-python3 robotframework
'';

File: pkgs/development/python-modules/robotframework/default.nix [hunk 7]
```diff
@@ -1,17 +1,18 @@
-{ stdenv, fetchurl, buildPythonPackage, isPy3k }:
+{ stdenv, fetchPypi, buildPythonPackage }:
+
+with stdenv.lib;
```
Comment (dotlambda) [inline pkgs/development/python-modules/robotframework/default.nix]: Why did you move that here?

File: pkgs/development/python-modules/robotsuite/default.nix (hunk 4 above)
Comment (rnhmjoj) [reply pkgs/development/python-modules/robotsuite/default.nix]: Ok, done.

File: pkgs/servers/mail/mailman/default.nix [hunk 8]
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, python, pythonPackages }:
+{ stdenv, fetchurl, python2, python2Packages }:
```
Comment (dotlambda) [inline pkgs/servers/mail/mailman/default.nix]: This should better only be `python2` and instead of `python2Packages` we can then use `python2.pkgs`.

File: pkgs/servers/matrix-synapse/default.nix [hunk 9]
```diff
@@ -1,6 +1,6 @@
-{ stdenv, pythonPackages, fetchurl, fetchFromGitHub }:
+{ stdenv, python2Packages, fetchurl, fetchFromGitHub, fetchpatch }:
```
Comment (dotlambda) [inline pkgs/servers/matrix-synapse/default.nix]: Why is `fetchpatch` added?

File: pkgs/servers/matrix-synapse/default.nix [hunk 10]
```diff
@@ -20,45 +20,45 @@ let
...
+  buildInputs = with python2Packages; [
     mock setuptoolsTrial
   ];
 
   meta = with stdenv.lib; {
     homepage = https://matrix.org;
     description = "Matrix reference homeserver";
     license = licenses.asl20;
-    maintainers = with maintainers; [ ralith roblabla ekleog ];
+    maintainers = [ maintainers.ralith maintainers.roblabla ];
```
Comment (dotlambda) [inline pkgs/servers/matrix-synapse/default.nix]: Why did you change this?

File: pkgs/servers/matrix-synapse/default.nix [hunk 11]
```diff
@@ -20,45 +20,45 @@ let
...
 
   patches = [
     ./matrix-synapse.patch
   ];
 
-  propagatedBuildInputs = with pythonPackages; [
+  propagatedBuildInputs = with python2Packages; [
     blist canonicaljson daemonize dateutil frozendict pillow pyasn1
     pydenticon pymacaroons-pynacl pynacl pyopenssl pysaml2 pytz requests
-    signedjson systemd twisted ujson unpaddedbase64 pyyaml prometheus_client
```
Comment (dotlambda) [inline pkgs/servers/matrix-synapse/default.nix]: I think `prometheus_client` was added on purpose: https://github.com/matrix-org/synapse/blob/1241156c82644d5609f45659607a356af5d8fe08/synapse/python_dependencies.py#L59
It's an optional dependecny however. Is there any policy regarding these, @FRidh?

File: pkgs/servers/matrix-synapse/default.nix [hunk 12]
```diff
@@ -20,45 +20,45 @@ let
...
+    signedjson systemd twisted ujson unpaddedbase64 pyyaml
     matrix-angular-sdk bleach netaddr jinja2 psycopg2
     psutil msgpack-python lxml matrix-synapse-ldap3
-    phonenumbers jsonschema affinity bcrypt sortedcontainers
```
Comment (dotlambda) [inline pkgs/servers/matrix-synapse/default.nix]: This one as well: https://github.com/matrix-org/synapse/blob/1241156c82644d5609f45659607a356af5d8fe08/synapse/python_dependencies.py#L53

File: pkgs/servers/varnish/default.nix [hunk 13]
```diff
@@ -1,5 +1,5 @@
 { stdenv, fetchurl, pcre, libxslt, groff, ncurses, pkgconfig, readline, libedit
-, python, pythonPackages, makeWrapper }:
+, python2, python2Packages, makeWrapper }:
```
Comment (dotlambda) [inline pkgs/servers/varnish/default.nix]: `python2` only

File: pkgs/servers/web-apps/searx/default.nix [hunk 14]
```diff
@@ -34,6 +34,9 @@ buildPythonApplication rec {
 
   checkInputs = [ splinter mock plone-testing robotsuite unittest2 ];
 
+  # test TestPubmedEngine is failing
+  doCheck = false;
```
Comment (dotlambda) [inline pkgs/servers/web-apps/searx/default.nix]: Might be better to just remove the single test in `postPatch`. However, I'm not able to test this because `searx` does not build.

**Comment by dotlambda** [comment]
@GrahamcOfBorg build searx

File: pkgs/servers/matrix-synapse/default.nix (hunk 11 above)
Comment (FRidh) [reply pkgs/servers/matrix-synapse/default.nix]: It's up to the maintainers to decide what to include and what not.

Comment (dotlambda) [reply pkgs/servers/matrix-synapse/default.nix]: Then we shouldn't remove it.

**Comment by dotlambda** [comment]
Please drop 744d9d0 from this PR, I've included it in https://github.com/NixOS/nixpkgs/pull/44398.

**Comment by rnhmjoj** [comment]
@dotlambda Ok, I removed the bluez commit.

File: pkgs/servers/matrix-synapse/default.nix (hunk 10 above)
Comment (rnhmjoj) [reply pkgs/servers/matrix-synapse/default.nix]: Sorry, I made this branch quite some time ago and I must have messed this up when rebasing to a recent nixpkgs master.

File: pkgs/servers/web-apps/searx/default.nix (hunk 14 above)
Comment (rnhmjoj) [reply pkgs/servers/web-apps/searx/default.nix]: Done. I'm not sure why the test is failing: lxml expects a string while the response is a mock object.

**Comment by rnhmjoj** [comment]
Thank you @dotlambda.
